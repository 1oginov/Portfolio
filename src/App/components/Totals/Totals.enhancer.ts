import { compose, withProps } from 'recompose';

import withCurrency from 'Currencies/enhancers/withCurrency';

import { Props } from './Totals';
import { calculateTotals, StockPosition } from '../../lib';

interface EnhancedProps {
  showClosed?: boolean;
  stockPositions: StockPosition[];
}

export default compose<Props, EnhancedProps>(
  withProps<Partial<Props>, EnhancedProps>(({ stockPositions }) => calculateTotals(stockPositions)),
  withCurrency,
);
