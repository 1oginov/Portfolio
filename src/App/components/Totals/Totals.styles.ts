import { withStyles } from '@material-ui/core';

export default withStyles(theme => ({
  group: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
  },
  root: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    padding: theme.spacing(2),
  },
  secondary: {
    display: 'flex',
  },
  sum: {
    marginBottom: theme.spacing(1),
    marginTop: theme.spacing(1),
  },
}));
