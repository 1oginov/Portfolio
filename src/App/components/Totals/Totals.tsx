import { Typography } from '@material-ui/core';
import * as React from 'react';

import Money from 'Shared/components/Money';
import Percent from 'Shared/components/Percent';

export interface Props {
  classes: { [key: string]: string };
  currency: string;
  currencyMultiplier: number | null;
  showClosed?: boolean;
  totalClosePL: number;
  totalClosePLPercent: number;
  totalCloseSum: number;
  totalDailyPL: number;
  totalDailyPLPercent: number;
  totalMarketPL: number;
  totalMarketPLPercent: number;
  totalMarketSum: number;
}

const Totals: React.FunctionComponent<Props> = ({
  classes, currency, currencyMultiplier, showClosed, totalClosePL, totalClosePLPercent, totalCloseSum, totalDailyPL,
  totalDailyPLPercent, totalMarketPL, totalMarketPLPercent, totalMarketSum,
}: Props) => {
  const sum = showClosed ? totalCloseSum : totalMarketSum;

  const groups = showClosed ? [
    <div className={classes.group} key="closed">
      <Money currency={currency} multiplier={currencyMultiplier} pl value={totalClosePL} />
      <Percent pl value={totalClosePLPercent} />
    </div>,
  ] : [
    <div className={classes.group} key="daily">
      <Money currency={currency} multiplier={currencyMultiplier} pl value={totalDailyPL} />
      <Percent pl value={totalDailyPLPercent} />
    </div>,
    <div className={classes.group} key="market">
      <Money currency={currency} multiplier={currencyMultiplier} pl value={totalMarketPL} />
      <Percent pl value={totalMarketPLPercent} />
    </div>,
  ];

  return (
    <div className={classes.root}>
      <Typography className={classes.sum} variant="h5">
        <Money currency={currency} multiplier={currencyMultiplier} value={sum} />
      </Typography>
      <div className={classes.secondary}>
        {groups}
      </div>
    </div>
  );
};

export default Totals;
