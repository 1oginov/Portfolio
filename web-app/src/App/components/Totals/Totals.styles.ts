import { withStyles } from '@material-ui/core';

export default withStyles(theme => ({
  currencyButton: {
    position: 'absolute',
    right: theme.spacing(0.5),
    top: theme.spacing(2),
  },
  group: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
  },
  root: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    padding: theme.spacing(2),
    position: 'relative',
  },
  secondary: {
    display: 'flex',
  },
  sum: {
    marginBottom: theme.spacing(1),
    marginTop: theme.spacing(1),
  },
}));
