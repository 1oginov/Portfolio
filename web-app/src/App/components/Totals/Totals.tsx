import {
  IconButton, Menu, MenuItem, Typography,
} from '@material-ui/core';
import { AccountBalanceWalletOutlined } from '@material-ui/icons';
import * as React from 'react';

import * as C from 'Constants';
// TODO: Should be unnecessary.
import {
  ChangeCurrencyAction,
} from 'Layers/Application/ActionCreators/CurrenciesActionCreators/CurrenciesActionCreators';
import { Money } from 'Layers/Presentation/Components/Money';
import { Percent } from 'Layers/Presentation/Components/Percent';

export interface Props {
  anchor: HTMLElement | null;
  changeCurrency: ChangeCurrencyAction;
  classes: { [key: string]: string };
  currency: string;
  currencyMultiplier: number | null;
  showClosed?: boolean;
  totalClosePL: number;
  totalClosePLPercent: number;
  totalCloseSum: number;
  totalDailyPL: number;
  totalDailyPLPercent: number;
  totalMarketPL: number;
  totalMarketPLPercent: number;
  totalMarketSum: number;
  updateAnchor: (anchor: HTMLElement | null) => void;
}

const Totals: React.FunctionComponent<Props> = ({
  anchor, changeCurrency, classes, currency, currencyMultiplier, showClosed, totalClosePL, totalClosePLPercent,
  totalCloseSum, totalDailyPL, totalDailyPLPercent, totalMarketPL, totalMarketPLPercent, totalMarketSum, updateAnchor,
}: Props) => {
  const sum = showClosed ? totalCloseSum : totalMarketSum;

  const groups = showClosed ? [
    <div className={classes.group} key="closed">
      <Money currency={currency} multiplier={currencyMultiplier} pl value={totalClosePL} />
      <Percent pl value={totalClosePLPercent} />
    </div>,
  ] : [
    <div className={classes.group} key="daily">
      <Money currency={currency} multiplier={currencyMultiplier} pl value={totalDailyPL} />
      <Percent pl value={totalDailyPLPercent} />
    </div>,
    <div className={classes.group} key="market">
      <Money currency={currency} multiplier={currencyMultiplier} pl value={totalMarketPL} />
      <Percent pl value={totalMarketPLPercent} />
    </div>,
  ];

  return (
    <div className={classes.root}>

      <Typography className={classes.sum} variant="h5">
        <Money currency={currency} multiplier={currencyMultiplier} value={sum} />
      </Typography>

      <div className={classes.secondary}>{groups}</div>

      <IconButton className={classes.currencyButton} onClick={(event) => updateAnchor(event.currentTarget)}>
        <AccountBalanceWalletOutlined />
      </IconButton>

      <Menu anchorEl={anchor} onClose={() => updateAnchor(null)} open={Boolean(anchor)}>
        {C.AVAILABLE_CURRENCIES.map(({ key, label }) => (
          <MenuItem
            disabled={key === currency}
            key={key}
            onClick={() => {
              changeCurrency(key);
              updateAnchor(null);
            }}
          >
            {label}
          </MenuItem>
        ))}
      </Menu>

    </div>
  );
};

export default Totals;
