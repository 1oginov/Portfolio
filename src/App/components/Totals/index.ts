import Totals from './Totals';
import enhance from './Totals.enhancer';
import style from './Totals.styles';

export default style(enhance(Totals));
