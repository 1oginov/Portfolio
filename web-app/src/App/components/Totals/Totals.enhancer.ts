import {
  compose, lifecycle, withProps, withState,
} from 'recompose';

import {
  CurrencyConnector, CurrencyConnectorProps,
} from 'Layers/Adapter/Connectors/CurrencyConnector/CurrencyConnector';

import { calculateTotals, StockPosition } from '../../lib';
import { Props } from './Totals';

interface EnhancedProps {
  showClosed?: boolean;
  stockPositions: StockPosition[];
}

export default compose<Props, EnhancedProps>(
  withState<EnhancedProps, HTMLElement | null, 'anchor', 'updateAnchor'>('anchor', 'updateAnchor', null),
  withProps<Partial<Props>, EnhancedProps>(({ stockPositions }) => calculateTotals(stockPositions)),
  CurrencyConnector,
  lifecycle<CurrencyConnectorProps, Record<string, never>>({

    componentDidMount() {
      this.props.fetchRates();
    },

  }),
);
