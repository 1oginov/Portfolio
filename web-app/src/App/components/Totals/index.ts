import { TotalsEnhancer } from 'App/components/Totals/TotalsEnhancer';

import { Totals as TotalsComponent, TotalsProps } from './Totals';
import { style } from './Totals.style';

export const Totals = style(TotalsEnhancer<TotalsProps>()(TotalsComponent));
