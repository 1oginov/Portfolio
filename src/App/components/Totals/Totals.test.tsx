import { shallow } from 'enzyme';
import classes from 'identity-obj-proxy';
import * as React from 'react';

import Totals from './Totals';

it('matches snapshot', () => {
  const wrapper = shallow(
    <Totals
      classes={classes}
      totalClosePL={100}
      totalClosePLPercent={1}
      totalCloseSum={1000}
      totalDailyPL={10}
      totalDailyPLPercent={0.1}
      totalMarketPL={100}
      totalMarketPLPercent={1}
      totalMarketSum={1000}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});

it('matches snapshot when showing closed', () => {
  const wrapper = shallow(
    <Totals
      classes={classes}
      showClosed
      totalClosePL={100}
      totalClosePLPercent={1}
      totalCloseSum={1000}
      totalDailyPL={10}
      totalDailyPLPercent={0.1}
      totalMarketPL={100}
      totalMarketPLPercent={1}
      totalMarketSum={1000}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
