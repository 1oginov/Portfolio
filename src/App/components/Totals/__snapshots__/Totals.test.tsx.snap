// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="root"
>
  <WithStyles(ForwardRef(Typography))
    className="sum"
    variant="h5"
  >
    <WithStyles(Money)
      currency="USD"
      multiplier={1}
      value={1000}
    />
  </WithStyles(ForwardRef(Typography))>
  <div
    className="secondary"
  >
    <div
      className="group"
      key="daily"
    >
      <WithStyles(Money)
        currency="USD"
        multiplier={1}
        pl={true}
        value={10}
      />
      <WithStyles(Percent)
        pl={true}
        value={0.1}
      />
    </div>
    <div
      className="group"
      key="market"
    >
      <WithStyles(Money)
        currency="USD"
        multiplier={1}
        pl={true}
        value={100}
      />
      <WithStyles(Percent)
        pl={true}
        value={1}
      />
    </div>
  </div>
</div>
`;

exports[`matches snapshot when showing closed 1`] = `
<div
  className="root"
>
  <WithStyles(ForwardRef(Typography))
    className="sum"
    variant="h5"
  >
    <WithStyles(Money)
      currency="USD"
      multiplier={1}
      value={1000}
    />
  </WithStyles(ForwardRef(Typography))>
  <div
    className="secondary"
  >
    <div
      className="group"
      key="closed"
    >
      <WithStyles(Money)
        currency="USD"
        multiplier={1}
        pl={true}
        value={100}
      />
      <WithStyles(Percent)
        pl={true}
        value={1}
      />
    </div>
  </div>
</div>
`;
